### PR TITLE
add CAAM with INTEGRITY support

### PR DESCRIPTION
**Pull request by JacobBarthelmeh**

### CODE REVIEW

**Comment by cconlon** [comment]
The PR from Go (https://github.com/wolfSSL/wolfssl/pull/1269) should fix the Jenkins enable options test that is failing.  Reviewing that PR now.

**Comment by JacobBarthelmeh** [comment]
Algorithms requiring attention with CAAM driver:
AES-CCM (partially complete)
AES-GCM (as much as possible)
AES XTS
SHA256 (done)

File: wolfcrypt/src/error.c [hunk 1]
```diff
@@ -434,6 +434,12 @@ const char* wc_GetErrorString(int error)
     case ECC_PRIVATEONLY_E:
         return "Invalid use of private only ECC key";
 
+    case WC_CAAM_E:
```
Comment (dgarske) [inline wolfcrypt/src/error.c]: Can we use a generic crypto hardware timeout / wait enum name here? Something like WC_HW_E and WC_HW_TIMEOUT_E?

File: wolfcrypt/src/aes.c [hunk 2]
```diff
@@ -2131,10 +2151,27 @@ static void wc_AesDecrypt(Aes* aes, const byte* inBlock, byte* outBlock)
...
     #if defined(AES_MAX_KEY_SIZE)
         const word32 max_key_len = (AES_MAX_KEY_SIZE / 8);
     #endif
 
+    #ifdef WOLFSSL_IMX6_CAAM_BLOB
+        byte   local[32];
+        word32 localSz = 32;
+
+        if (keylen == (16 + WC_CAAM_BLOB_SZ) ||
+         keylen == (24 + WC_CAAM_BLOB_SZ) ||
```
Comment (dgarske) [inline wolfcrypt/src/aes.c]: Check indentation here (tab vs. spaces).

File: wolfcrypt/src/aes.c [hunk 3]
```diff
@@ -2178,15 +2215,45 @@ static void wc_AesDecrypt(Aes* aes, const byte* inBlock, byte* outBlock)
         }
     #endif /* WOLFSSL_AESNI */
 
-        return wc_AesSetKeyLocal(aes, userKey, keylen, iv, dir);
+        ret = wc_AesSetKeyLocal(aes, userKey, keylen, iv, dir);
+
+    #ifdef WOLFSSL_IMX6_CAAM_BLOB
+	ForceZero(local, sizeof(local));
```
Comment (dgarske) [inline wolfcrypt/src/aes.c]: Check indentation here (tab vs. spaces).

File: wolfcrypt/src/aes.c [hunk 4]
```diff
@@ -2178,15 +2215,45 @@ static void wc_AesDecrypt(Aes* aes, const byte* inBlock, byte* outBlock)
...
+
+                /* set local values */
+                userKey = local;
+                keylen = localSz;
+            }
+        #endif
+            ret = wc_AesSetKeyLocal(aes, userKey, keylen, iv, dir);
+
+        #ifdef WOLFSSL_IMX6_CAAM_BLOB
+	        ForceZero(local, sizeof(local));
```
Comment (dgarske) [inline wolfcrypt/src/aes.c]: Check indentation here (tab vs. spaces).

File: wolfcrypt/src/port/caam/caam_aes.c [hunk 5]
```diff
@@ -0,0 +1,648 @@
+/* caam_aes.c
+ *
+ * Copyright (C) 2006-2016 wolfSSL Inc.
```
Comment (dgarske) [inline wolfcrypt/src/port/caam/caam_aes.c]: Copyright 2017. Update in all new files as well.

File: wolfcrypt/src/port/caam/caam_driver.c [hunk 6]
```diff
@@ -0,0 +1,1709 @@
...
+/******************************************************************************
+  Internal CAAM Job Ring and partition functions
+  ****************************************************************************/
+
+/* flush job ring and reset */
+static Error caamReset()
+{
+    int t = 100000; /* time out counter for flushing job ring */
+
+    /* make sure interupts are masked in JRCFGR0_LS register */
```
Comment (dgarske) [inline wolfcrypt/src/port/caam/caam_driver.c]: Spelling "interrupts"

File: wolfcrypt/src/port/caam/caam_driver.c [hunk 7]
```diff
@@ -0,0 +1,1709 @@
...
+    CAAM_WRITE(CAAM_BASE | 0x1054, CAAM_READ(CAAM_BASE | 0x1054) | 1);
+
+    /* flush and reset job rings using JRCR0 register */
+    CAAM_WRITE(CAAM_BASE | 0x106C, 1);
+
+    /* check register JRINTR for if halt is in prgress */
```
Comment (dgarske) [inline wolfcrypt/src/port/caam/caam_driver.c]: Spelling "progress"

File: wolfcrypt/src/port/caam/caam_driver.c [hunk 8]
```diff
@@ -0,0 +1,1709 @@
...
+ * all other returns is an error state
+ */
+static Error caamCreatePartition(unsigned char page, unsigned char par)
+{
+    /* check ownership of partition */
+    if ((CAAM_READ(CAAM_BASE | 0x1FBC) & (0x3 << (par * 2))) > 0) {
+        return MemoryMapMayNotBeEmpty;
+    }
+
+    /* set generic all access permisions, gets reset later */
```
Comment (dgarske) [inline wolfcrypt/src/port/caam/caam_driver.c]: Spelling "permissions"

File: wolfcrypt/src/port/caam/caam_driver.c [hunk 9]
```diff
@@ -0,0 +1,1709 @@
...
+    CAAM_WRITE(CAAM_BASE | (0x1108 + (par * 16)), 0xF);
+    CAAM_WRITE(CAAM_BASE | (0x110C + (par * 16)), 0xF);
+    CAAM_WRITE(CAAM_BASE | (0x1104 + (par * 16)), 0xFF);
+
+    /* check ownership of page */
+    CAAM_WRITE(CAAM_BASE | 0x10F4, (page << 16) | 0x5);
+    /* wait for inquery cmd to complete */
```
Comment (dgarske) [inline wolfcrypt/src/port/caam/caam_driver.c]: Spelling "inquiry"

File: wolfcrypt/src/port/caam/caam_driver.c [hunk 10]
```diff
@@ -0,0 +1,1709 @@
...
+       (CAAM_READ(CAAM_BASE | 0x10FC) & 0x00003000)  == 0) {
+    }
+    if ((CAAM_READ(CAAM_BASE | 0x10FC) & 0x000000C0) == 0xC0) {
+        /* owns the page can dealloc it */
+        CAAM_WRITE(CAAM_BASE | 0x10F4, (page << 16) | 0x2);
+        while ((CAAM_READ(CAAM_BASE | 0x10FC) & 0x0000C000) > 0 &&
+               (CAAM_READ(CAAM_BASE | 0x10FC) & 0x00003000)  == 0) {}
+        if ((CAAM_READ(CAAM_BASE | 0x10FC) & 0x00003000)  > 0) {
+            /* error while deallocating page */
+            return MemoryMapMayNotBeEmpty; /* PSP set on page or is unavialbale */
```
Comment (dgarske) [inline wolfcrypt/src/port/caam/caam_driver.c]: Spelling "unavailable"

File: wolfcrypt/src/port/caam/caam_driver.c [hunk 11]
```diff
@@ -0,0 +1,1709 @@
...
+       (CAAM_READ(CAAM_BASE | 0x10FC) & 0x00003000)  == 0) {
+    }
+
+    if ((CAAM_READ(CAAM_BASE | 0x10FC) & 0x00003000) > 0) {
+        return MemoryOperationNotPerformed;
+    }
+
+    /* double check ownership now of page */
+    CAAM_WRITE(CAAM_BASE | 0x10F4, (page << 16) | 0x5);
+    /* wait for inquery cmd to complete */
```
Comment (dgarske) [inline wolfcrypt/src/port/caam/caam_driver.c]: Spelling "inquiry"

File: wolfcrypt/src/port/caam/caam_driver.c [hunk 12]
```diff
@@ -0,0 +1,1709 @@
...
+
+    /* Set back to run mode and clear RTMCL error bit */
+    reg = CAAM_READ(CAAM_RTMCTL) ^ CAAM_PRGM;
+
+    CAAM_WRITE(CAAM_RTMCTL, reg);
+    reg = CAAM_READ(CAAM_RTMCTL);
+    reg |= CAAM_CTLERR;
+    CAAM_WRITE(CAAM_RTMCTL, reg);
+
+    /* check input slot is avialable and then add */
```
Comment (dgarske) [inline wolfcrypt/src/port/caam/caam_driver.c]: Spelling "available"

File: wolfcrypt/src/port/caam/caam_driver.c [hunk 13]
```diff
@@ -0,0 +1,1709 @@
...
+static Error caamDoJob(struct DescStruct* desc)
+{
+    Error ret;
+    UINT4 status;
+
+    /* clear and set desc size */
+    desc->desc[0] &= 0xFFFFFF80;
+    desc->desc[0] += desc->idx;
+
+    /* check input slot is avialable and then add */
```
Comment (dgarske) [inline wolfcrypt/src/port/caam/caam_driver.c]: Spelling "available"

File: wolfcrypt/src/port/caam/caam_driver.c [hunk 14]
```diff
@@ -0,0 +1,1709 @@
...
+    if (CAAM_READ(CAAM_BASE | 0x1014) > 0) {
+        UINT4* in = (UINT4*)desc->caam->ring.JobIn;
+
+        memcpy((unsigned char*)desc->caam->ring.Desc, (unsigned char*)desc->desc,
```
Comment (dgarske) [inline wolfcrypt/src/port/caam/caam_driver.c]: Should we use XMEMCPY and XMEMSET where applicable?

File: wolfcrypt/src/port/caam/caam_driver.c [hunk 15]
```diff
@@ -0,0 +1,1709 @@
...
+        int blocks = buf->dataSz / align;
+        Address data   = buf->data;
+        Address dataSz = buf->dataSz;
+
+        if (outSz >= sz && sz != 0) {
+            break;
+        }
+
+        if (dataSz % align > 0) {
+            /* store potental overlap */
```
Comment (dgarske) [inline wolfcrypt/src/port/caam/caam_driver.c]: Spelling "potential"

File: wolfcrypt/src/port/caam/caam_driver.c [hunk 16]
```diff
@@ -0,0 +1,1709 @@
...
+
+    return Success;
+}
+
+
+/******************************************************************************
+  CAAM AES Operations
+  ****************************************************************************/
+
+/* returns amount writin on success and negative value in error case.
```
Comment (dgarske) [inline wolfcrypt/src/port/caam/caam_driver.c]: Spelling "written"

File: wolfcrypt/src/port/caam/caam_driver.c [hunk 17]
```diff
@@ -0,0 +1,1709 @@
...
+    desc->desc[desc->idx++] = BSP_VirtualToPhysical(desc->ctxBuf);
+
+    /* get IV if needed by algorithm */
+    switch (desc->type) {
+        case CAAM_AESECB:
+            break;
+
+        case CAAM_AESCTR:
+            ofst = 0x00001000;
+            /* fall through because states are the same only the offest changes */
```
Comment (dgarske) [inline wolfcrypt/src/port/caam/caam_driver.c]: Spelling "offset"

File: wolfcrypt/src/port/caam/caam_driver.c [hunk 18]
```diff
@@ -0,0 +1,1709 @@
...
+        /********* handle AAD -- is only done with Init **********************/
+        if ((state & CAAM_ALG_INIT) == CAAM_ALG_INIT) {
+            if ((desc->type == CAAM_AESCCM) && (desc->aadSz > 0)) {
+                /* set formated AAD buffer size for CCM */
+                ASP_FlushCaches((Address)desc->aadSzBuf, sizeof(desc->aadSzBuf));
+                desc->desc[desc->idx++] = CAAM_FIFO_L | CAAM_CLASS1 |
+                    FIFOL_TYPE_AAD + desc->aadSz;
+                desc->desc[desc->idx++] = BSP_VirtualToPhysical(desc->aadSzBuf);
+
+                /* now set aadSz to unformated version for getting buffers */
```
Comment (dgarske) [inline wolfcrypt/src/port/caam/caam_driver.c]: Spelling "unformatted"

File: wolfcrypt/src/port/caam/caam_driver.c [hunk 19]
```diff
@@ -0,0 +1,1709 @@
...
+    if (caamAesOutSz(desc, i) != Success) {
+        return SizeIsTooLarge;
+    }
+
+    /* set alignment constraints */
+    if (desc->type == CAAM_AESCBC || desc->type == CAAM_AESECB) {
+        align = 16;
+    }
+
+    /* indefinit loop for input/output buffers */
```
Comment (dgarske) [inline wolfcrypt/src/port/caam/caam_driver.c]: Spelling "indefinite"

File: wolfcrypt/src/port/caam/caam_driver.c [hunk 20]
```diff
@@ -0,0 +1,1709 @@
...
+        }
+    }
+    if (sz > ctxSz || ctxSz > (MAX_CTX * sizeof(UINT4))) {
+        return SizeIsTooLarge;
+    }
+
+    ASP_FlushCaches((Address)desc->iv, ctxSz);
+    /*Manage Context (current digest + 8 byte running message length)*/
+    if ((desc->state & CAAM_ALG_INIT) != CAAM_ALG_INIT) {
+        /* dont load into the class 2 context register on inti.
```
Comment (dgarske) [inline wolfcrypt/src/port/caam/caam_driver.c]: Grammar "don't"

File: wolfcrypt/src/port/caam/caam_driver.c [hunk 21]
```diff
@@ -0,0 +1,1709 @@
...
+}
+
+
+static void HandleInterrupt(Address id)
+{
+    struct CAAM_DEVICE* local = (struct CAAM_DEVICE*)id;
+    Value InterruptStatus = INTERRUPT_AtomicWrite(&local->InterruptStatus, 0);
+    int i;
+
+    /* Loop through decriptors and try to dismiss them */
```
Comment (dgarske) [inline wolfcrypt/src/port/caam/caam_driver.c]: Spelling "descriptors"

File: wolfcrypt/src/port/caam/caam_driver.c [hunk 22]
```diff
@@ -0,0 +1,1709 @@
...
+    /* Initialize Descriptors */
+    for (i = 0; i < DESC_COUNT; i++) {
+         InitializeIORequest(ioCaam, &caam.DescArray[i].TheIORequest,
+                             IOREQUEST_STANDARD);
+         caam.DescArray[i].running = false;
+         caam.DescArray[i].caam    = &caam;
+    }
+
+
+    /* call interupt to make IORequests available */
```
Comment (dgarske) [inline wolfcrypt/src/port/caam/caam_driver.c]: Spelling "interrupt"

File: wolfcrypt/src/port/caam/caam_driver.c [hunk 23]
```diff
@@ -0,0 +1,1709 @@
...
+    #define HW_CCM_CCGR0_ADDR (0xf20c4068)
+    #define CG(x) (3 << (x*2))
+
+    reg = CG(6) | CG(5) | CG(4);
+    *(volatile unsigned int*)HW_CCM_CCGR0_ADDR =
+    *(volatile unsigned int*)HW_CCM_CCGR0_ADDR | reg;
+
+    /* set up job ring */
+
+    /* @TODO create partion in physical memory for job rings
```
Comment (dgarske) [inline wolfcrypt/src/port/caam/caam_driver.c]: Spelling "partition"

File: wolfcrypt/src/port/caam/caam_driver.c (hunk 14 above)
Comment (JacobBarthelmeh) [reply wolfcrypt/src/port/caam/caam_driver.c]: I would agree except that the file caam_driver.c is to be compiled with the kernel separate from wolfSSL. The kernel gets two files caam_driver.c and caam_driver.h, maybe I should add a README.md to this directory with that information to help people not compile caam_driver.c in their application?

**Comment by JacobBarthelmeh** [comment]
Thanks for the review David, am going through and fixing the spelling and formatting errors.

**Pull request merged**